circus-train-api/src/main/java/com/hotels/bdp/circustrain/api/metastore/compatibility/HiveMetaStoreClientCompatibility12x.java

```diff
@@ -0,0 +1,128 @@
+/**
+ * Copyright (C) 2016-2018 Expedia Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package com.hotels.bdp.circustrain.api.metastore.compatibility;
+
+import java.lang.reflect.Field;
+import java.lang.reflect.InvocationHandler;
+import java.lang.reflect.InvocationTargetException;
+import java.lang.reflect.Method;
+import java.lang.reflect.Proxy;
+
+import org.apache.hadoop.hive.metastore.IMetaStoreClient;
+import org.apache.hadoop.hive.metastore.RetryingMetaStoreClient;
+import org.apache.hadoop.hive.metastore.api.MetaException;
+import org.apache.hadoop.hive.metastore.api.NoSuchObjectException;
+import org.apache.hadoop.hive.metastore.api.Table;
+import org.apache.hadoop.hive.metastore.api.ThriftHiveMetastore;
+import org.apache.thrift.TApplicationException;
+import org.apache.thrift.TBase;
+import org.apache.thrift.TException;
+import org.apache.thrift.TServiceClient;
+
+public class HiveMetaStoreClientCompatibility12x {
+
+  private final TServiceClient tServiceClient;
+
+  public HiveMetaStoreClientCompatibility12x(IMetaStoreClient client) {
+    while (Proxy.isProxyClass(client.getClass())) {
+      InvocationHandler handler = Proxy.getInvocationHandler(client);
+      if (handler.getClass().isAssignableFrom(RetryingMetaStoreClient.class)) {
+        client = getField(handler, "base");
+        continue;
+      }
+      // TODO add CloseableMetaStoreClientInvocationHandler ?
+      throw new RuntimeException("Unknown InvocationHandler " + handler.getClass());
+    }
+    tServiceClient = getField(client, "client");
+  }
+
+  private static <T> T getField(Object object, String fieldName) {
+    try {
+      Field field = object.getClass().getDeclaredField(fieldName);
+      T result = null;
+      if (field.isAccessible()) {
+        result = (T) field.get(object);
+      } else {
+        field.setAccessible(true);
+        result = (T) field.get(object);
+        field.setAccessible(false);
+      }
+      return result;
+    } catch (Exception e) {
+      throw new RuntimeException("Unable to hack client", e);
```

Comment:
Bonus points for amusing exception message

Reply:
:) if you have a better message let me know